community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/SchemaCache.java

```diff
@@ -40,209 +42,279 @@
 import org.neo4j.storageengine.api.schema.SchemaRule;
 
 import static java.util.Collections.emptyIterator;
-import static java.util.Collections.newSetFromMap;
 
 /**
  * A cache of {@link SchemaRule schema rules} as well as enforcement of schema consistency.
  * Will always reflect the committed state of the schema store.
- *
- * Assume synchronization/locking is done outside, with locks.
- *
- * @author Mattias Persson
- * @author Stefan Plantikow
  */
 public class SchemaCache
 {
-    private final Map<Long,IndexRule> indexRuleById = new ConcurrentHashMap<>();
-    private final Map<Long,ConstraintRule> constraintRuleById = new ConcurrentHashMap<>();
-    private final Set<ConstraintDescriptor> constraints = newSetFromMap( new ConcurrentHashMap<>() );
-
-    private final Map<SchemaDescriptor,IndexDescriptor> indexDescriptors = new ConcurrentHashMap<>();
-    private final Map<Integer,Set<IndexDescriptor>> indexDescriptorsByLabel = new ConcurrentHashMap<>();
-    private final ConstraintSemantics constraintSemantics;
-
-    private final Map<Class<?>,Object> dependantState = new ConcurrentHashMap<>();
-    private final Map<Integer,List<IndexDescriptor>> indexByProperty = new ConcurrentHashMap<>();
+    private final Lock cacheUpdateLock = new StampedLock().asWriteLock();
```

Comment:
Why this particular lock instead of, say, `new ReentrantLock()`?

Reply:
Since we do not need reentrancy part and need simple single entry lock, decided to go for this one. But it can be any lock or semaphore in principle